src/Polysemy/Final.hs

```diff
@@ -174,24 +169,24 @@ withStrategicToFinal strat = withWeavingToFinal $ \s wv ins ->
 -- interpreted using 'interpretFinal'. See 'Final'.
 interpretFinal
     :: forall m e r a
-     . (Member (Final m) r, Functor m)
+     . Member (Final m) r
     => (forall x n. e n x -> Strategic m n x)
        -- ^ A natural transformation from the handled effect to the final monad.
     -> Sem (e ': r) a
     -> Sem r a
 interpretFinal n =
   let
     go :: Sem (e ': r) x -> Sem r x
-    go = usingSem $ \u -> case decomp u of
+    go = hoistSem $ \u -> case decomp u of
       Right (Weaving e s wv ex ins) ->
-        fmap ex $ withWeavingToFinal $ \s' wv' ins'
-          -> fmap getCompose $
-                runStrategy
-                  (Compose (s <$ s'))
-                  (fmap Compose . wv' . fmap (go . wv) . getCompose)
-                  (ins' . getCompose >=> ins)
-                  (n e)
-      Left g -> liftSem (hoist go g)
+        injWeaving $
+          Weaving
+            (WithWeavingToFinal (runStrategy (n e)))
+            s
+            (go . wv)
+            ex
+            ins
+      Left g -> hoist go g
```

Comment:
Found this neat optimisation made possible from the new `injWeaving`, which cleans up the code to boot!